modules/javafx.graphics/src/main/java/javafx/css/TransitionEvent.java

```diff
@@ -0,0 +1,117 @@
+import javafx.scene.Node;
+import javafx.util.Duration;
+
+/**
+ * An event that signals the creation, beginning, completion and cancellation of implicit CSS transitions.
+ * <p>
+ * Note that this event is not raised for explicit transitions that are created using the {@link Transition} class.
+ *
+ * @since 22
+ */
+public final class TransitionEvent extends Event {
+
+    private static final long serialVersionUID = 20220820L;
+
+    /**
+     * Common supertype for all {@code TransitionEvent} types.
+     */
+    public static final EventType<TransitionEvent> ANY =
+            new EventType<>(Event.ANY, "TRANSITION");
+
+    /**
+     * This event occurs when a transition has been created and added to the list of running
+     * transitions of a {@link Node}.
+     */
+    public static final EventType<TransitionEvent> RUN =
+            new EventType<>(TransitionEvent.ANY, "TRANSITION_RUN");
+
+    /**
+     * This event occurs when a running transition enters its active period, which happens
+     * at the end of the delay phase.
+     */
+    public static final EventType<TransitionEvent> START =
+            new EventType<>(TransitionEvent.ANY, "TRANSITION_START");
+
+    /**
+     * This event occurs when a running transition has reached the end of its active period.
+     * If the transition is cancelled prior to reaching the end of its active period, this
+     * event does not occur.
+     */
+    public static final EventType<TransitionEvent> END =
+            new EventType<>(TransitionEvent.ANY, "TRANSITION_END");
+
+    /**
+     * This event occurs when a running transition was cancelled before it has reached the
+     * end of its active period.
+     */
+    public static final EventType<TransitionEvent> CANCEL =
+            new EventType<>(TransitionEvent.ANY, "TRANSITION_CANCEL");
+
+    private final StyleableProperty<?> property;
+    private final Duration elapsedTime;
+
+    /**
+     * Creates a new instance of the {@code TransitionEvent} class.
+     *
+     * @param eventType the event type
+     * @param property the {@code StyleableProperty} that is targeted by the transition
+     * @param elapsedTime the time that has elapsed since the transition has entered its active period
+     */
+    public TransitionEvent(EventType<? extends Event> eventType,
+                           StyleableProperty<?> property,
+                           Duration elapsedTime) {
+        super(eventType);
+        this.property = property;
+        this.elapsedTime = elapsedTime;
+    }
+
+    /**
+     * Gets the {@code StyleableProperty} that is targeted by the transition.
+     *
+     * @return the {@code StyleableProperty}
+     */
+    public StyleableProperty<?> getProperty() {
+        return property;
+    }
+
+    /**
+     * Gets the time that has elapsed since the transition has entered its active period,
+     * not including the time spent in the delay phase.
+     *
+     * @return the elapsed time
```

Comment:
Any guarantees here?  Can it be `null`, negative, zero, infinite?

Reply:
Added that the elapsed time is zero if the transition has not entered its active period. I usually don't document that a return value is always non-null, since I believe this to be the default assumption unless otherwise noted. Additionally, the constructor of `TransitionEvent` now documents that it will not accept `null`.